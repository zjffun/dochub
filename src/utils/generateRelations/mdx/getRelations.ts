import { IRelation } from "../../../types.d";
import getBlocks from "./getBlocks";

// remark-frontmatter preset
const frontmatterType = ["yaml", "toml"];

export default async function getRelations(
  fromContent: string,
  toContent: string
): Promise<IRelation[]> {
  const { default: remarkFrontmatter } = await import("remark-frontmatter");
  const { default: remarkParse } = await import("remark-parse");
  const { unified } = await import("unified");

  const fromAst = await unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .parse(fromContent);
  const toAst = await unified()
    .use(remarkParse)
    .use(remarkFrontmatter)
    .parse(toContent);

  const fromBlocks = getBlocks(fromAst);
  const toBlocks = getBlocks(toAst);

  const relations: IRelation[] = [];

  let fromBlockIndex = 0;
  let toBlockIndex = 0;

  while (fromBlockIndex < fromBlocks.length && toBlockIndex < toBlocks.length) {
    const fromBlock = fromBlocks[fromBlockIndex];
    const toBlock = toBlocks[toBlockIndex];

    if (
      frontmatterType.includes(fromBlock.type || "") &&
      !frontmatterType.includes(toBlock.type || "")
    ) {
      fromBlockIndex++;
      continue;
    }

    if (
      frontmatterType.includes(toBlock.type || "") &&
      !frontmatterType.includes(fromBlock.type || "")
    ) {
      toBlockIndex++;
      continue;
    }

    relations.push({
      fromRange: [fromBlock.start, fromBlock.end],
      toRange: [toBlock.start, toBlock.end],
    });

    fromBlockIndex++;
    toBlockIndex++;
  }

  return relations;
}
