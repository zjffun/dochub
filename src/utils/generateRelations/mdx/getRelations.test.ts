import getRelations from "./getRelations";

const baseFromContent = `# test 1

foo

## test 2

bar`;
const baseToContent = `# 测试 1

文字1
文字2

## 测试 2

文字3`;

describe("getRelations", () => {
  test("base", async () => {
    expect(
      await getRelations(baseFromContent, baseToContent)
    ).toMatchSnapshot();
  });

  test("frontmatter", async () => {
    const fromContent = `---
title: 中文标题
---

${baseFromContent}
`;

    const toContent = `---
title: 中文标题
---

${baseToContent}
`;

    expect(await getRelations(fromContent, toContent)).toMatchSnapshot();
  });

  test("frontmatter only in to content", async () => {
    const fromContent = `${baseFromContent}`;

    const toContent = `---
title: 中文标题
---

${baseToContent}
`;

    expect(await getRelations(fromContent, toContent)).toMatchSnapshot();
  });

  test("frontmatter only in from content", async () => {
    const fromContent = `---
title: 中文标题
---

${baseFromContent}
`;

    const toContent = `${baseToContent}`;

    expect(await getRelations(fromContent, toContent)).toMatchSnapshot();
  });
});
