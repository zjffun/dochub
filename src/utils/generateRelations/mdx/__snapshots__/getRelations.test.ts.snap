// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRelations base 1`] = `
Array [
  Object {
    "fromRange": Array [
      1,
      3,
    ],
    "toRange": Array [
      1,
      4,
    ],
  },
  Object {
    "fromRange": Array [
      5,
      7,
    ],
    "toRange": Array [
      6,
      8,
    ],
  },
]
`;

exports[`getRelations frontmatter 1`] = `
Array [
  Object {
    "fromRange": Array [
      1,
      3,
    ],
    "toRange": Array [
      1,
      3,
    ],
  },
  Object {
    "fromRange": Array [
      5,
      7,
    ],
    "toRange": Array [
      5,
      8,
    ],
  },
  Object {
    "fromRange": Array [
      9,
      11,
    ],
    "toRange": Array [
      10,
      12,
    ],
  },
]
`;

exports[`getRelations frontmatter only in from content 1`] = `
Array [
  Object {
    "fromRange": Array [
      5,
      7,
    ],
    "toRange": Array [
      1,
      4,
    ],
  },
  Object {
    "fromRange": Array [
      9,
      11,
    ],
    "toRange": Array [
      6,
      8,
    ],
  },
]
`;

exports[`getRelations frontmatter only in to content 1`] = `
Array [
  Object {
    "fromRange": Array [
      1,
      3,
    ],
    "toRange": Array [
      5,
      8,
    ],
  },
  Object {
    "fromRange": Array [
      5,
      7,
    ],
    "toRange": Array [
      10,
      12,
    ],
  },
]
`;
